File: packages/skill-eval-python/src/skill_eval/grader/prompts.py
from __future__ import annotations

import json

from skill_eval.runner.transcript import AgentRun, TranscriptStep
from skill_eval.schemas.evals import EvalCase, get_turns

MAX_GRADING_CHARS = 300_000
CAP_USER_MESSAGE = 1_000
CAP_ASSISTANT_RESPONSE = 3_000
CAP_OUTPUT_FILE = 5_000
CAP_TOOL_VALUE = 2_000
CAP_SKIPPED_TURN = 200


def _cap(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    return f"{text[:limit]}\n\n... [truncated {len(text) - limit} chars] ..."


def _trunc_val(val: object, max_len: int = CAP_TOOL_VALUE) -> str:
    s = val if isinstance(val, str) else json.dumps(val, default=str)
    return f"{s[:max_len]}… [{len(s) - max_len} chars truncated]" if len(s) > max_len else s


def _format_transcript_step(
    step: TranscriptStep, last_turn: int
) -> tuple[str, int]:
    parts: list[str] = []
    current_last_turn = last_turn

    if step.turn is not None and step.turn != current_last_turn:
        current_last_turn = step.turn
        if step.turn > 0:
            parts.append(f"\n=== User Turn {step.turn + 1} ===")
            if step.user_message:
                parts.append(f"User: {_cap(step.user_message, CAP_USER_MESSAGE)}")
            parts.append("")

    parts.append(f"--- Step {step.step} ---")
    if step.response:
        parts.append(f"Assistant: {_cap(step.response, CAP_ASSISTANT_RESPONSE)}")
    for tc in step.tool_calls or []:
        parts.append(f"Tool call: {tc.name}({_trunc_val(tc.arguments)})")
    for tr in step.tool_results:
        parts.append(f"Tool result [{tr.name}]: {_trunc_val(tr.result)}")
    if step.turn_skipped:
        parts.append(
            f"\n⚠ TURN SKIPPED: The next scripted user message (\"{_cap(step.turn_skipped, CAP_SKIPPED_TURN)}\") "
            "was not injected because it did not match the agent's response. The conversation ended here."
        )

    return "\n".join(parts), current_last_turn


def _format_transcript(agent_run: AgentRun) -> str:
    last_turn = -1
    result: list[str] = []
    for step in agent_run.transcript:
        text, last_turn = _format_transcript_step(step, last_turn)
        result.append(text)
    return "\n\n".join(result)


def _format_output_files(
    output_files: list[dict[str, str]],
) -> str:
    if not output_files:
        return "(no output files)"
    return "\n\n".join(
        f"--- {f['path']} ---\n{_cap(f['content'], CAP_OUTPUT_FILE)}"
        for f in output_files
    )


def _format_turns_section(turns: list[str], is_multi_turn: bool) -> str:
    if is_multi_turn:
        lines = "\n".join(f"{i + 1}. {t}" for i, t in enumerate(turns))
        return (
            f"**Conversation turns:**\n{lines}\n\n"
            "This is a multi-turn conversation. The user sends the first message, "
            "the agent responds (possibly with tool calls), then the next user message "
            "is injected, and so on. Evaluate assertions across the full conversation."
        )
    return f"**Prompt given to the agent:**\n{turns[0]}"


def build_grading_prompt(
    eval_case: EvalCase,
    agent_run: AgentRun,
    output_files: list[dict[str, str]],
) -> str:
    turns = get_turns(eval_case)
    is_multi_turn = len(turns) > 1
    turns_section = _format_turns_section(turns, is_multi_turn)
    transcript_text = _format_transcript(agent_run)
    files_text = _format_output_files(output_files)
    assertions_list = "\n".join(f"{i + 1}. {a}" for i, a in enumerate(eval_case.assertions))

    prompt = f"""You are an evaluator grading the output of an AI agent that was given a task.

## Original Task

{turns_section}

**Expected behavior:**
{eval_case.expected_output}

## Assertions to Evaluate

Evaluate each of the following assertions. For each one, determine whether the agent's behavior satisfies it, and provide specific evidence from the transcript.

{assertions_list}

## Agent Transcript

{transcript_text}

## Output Files Produced

{files_text}

## Instructions

Return a JSON object with this exact structure (no other text, just JSON):

{{
  "expectations": [
    {{
      "text": "<the assertion text>",
      "passed": true/false,
      "evidence": "<specific evidence from the transcript>"
    }}
  ],
  "claims": [],
  "eval_feedback": "<optional overall feedback or null>"
}}

There must be exactly one entry in "expectations" for each assertion listed above, in the same order. Be strict: only mark an assertion as passed if the transcript clearly demonstrates it."""

    return _cap(prompt, MAX_GRADING_CHARS)


GRADER_SYSTEM_PROMPT = (
    "You are a precise evaluator. You grade AI agent outputs against specific assertions. "
    "Return only valid JSON matching the requested schema. Be strict but fair."
)

File: packages/skill-eval/src/grader/prompts.ts
import type { AgentRun } from "../runner/transcript.js";
import type { TranscriptStep } from "../runner/transcript.js";
import type { EvalCase } from "../schemas/evals.js";
import { getTurns } from "../schemas/evals.js";

const MAX_GRADING_CHARS = 300_000;
const CAP_USER_MESSAGE = 1_000;
const CAP_ASSISTANT_RESPONSE = 3_000;
const CAP_OUTPUT_FILE = 5_000;
const CAP_TOOL_VALUE = 2_000;
const CAP_SKIPPED_TURN = 200;

function cap(text: string, limit: number): string {
	if (text.length <= limit) return text;
	return `${text.slice(0, limit)}\n\n... [truncated ${text.length - limit} chars] ...`;
}

function truncVal(val: unknown, max = CAP_TOOL_VALUE): string {
	const s = typeof val === "string" ? val : JSON.stringify(val);
	return s.length > max ? `${s.slice(0, max)}… [${s.length - max} chars truncated]` : s;
}

function formatTranscriptStep(
	step: TranscriptStep,
	lastTurn: number,
): { text: string; newLastTurn: number } {
	const parts: string[] = [];
	let currentLastTurn = lastTurn;

	if (step.turn !== undefined && step.turn !== currentLastTurn) {
		currentLastTurn = step.turn;
		if (step.turn > 0) {
			parts.push(`\n=== User Turn ${step.turn + 1} ===`);
			if (step.userMessage) {
				parts.push(`User: ${cap(step.userMessage, CAP_USER_MESSAGE)}`);
			}
			parts.push("");
		}
	}

	parts.push(`--- Step ${step.step} ---`);
	if (step.response) parts.push(`Assistant: ${cap(step.response, CAP_ASSISTANT_RESPONSE)}`);
	for (const tc of step.toolCalls ?? []) {
		parts.push(`Tool call: ${tc.name}(${truncVal(tc.arguments)})`);
	}
	for (const tr of step.toolResults) {
		parts.push(`Tool result [${tr.name}]: ${truncVal(tr.result)}`);
	}
	if (step.turnSkipped) {
		parts.push(
			`\n⚠ TURN SKIPPED: The next scripted user message ("${cap(step.turnSkipped, CAP_SKIPPED_TURN)}") was not injected because it did not match the agent's response. The conversation ended here.`,
		);
	}

	return { text: parts.join("\n"), newLastTurn: currentLastTurn };
}

function formatTranscript(agentRun: AgentRun): string {
	let lastTurn = -1;
	return agentRun.transcript
		.map((step) => {
			const { text, newLastTurn } = formatTranscriptStep(step, lastTurn);
			lastTurn = newLastTurn;
			return text;
		})
		.join("\n\n");
}

function formatOutputFiles(outputFiles: Array<{ path: string; content: string }>): string {
	if (outputFiles.length === 0) return "(no output files)";
	return outputFiles
		.map((f) => `--- ${f.path} ---\n${cap(f.content, CAP_OUTPUT_FILE)}`)
		.join("\n\n");
}

function formatTurnsSection(turns: string[], isMultiTurn: boolean): string {
	if (isMultiTurn) {
		return `**Conversation turns:**\n${turns.map((t, i) => `${i + 1}. ${t}`).join("\n")}\n\nThis is a multi-turn conversation. The user sends the first message, the agent responds (possibly with tool calls), then the next user message is injected, and so on. Evaluate assertions across the full conversation.`;
	}
	return `**Prompt given to the agent:**\n${turns[0]}`;
}

export function buildGradingPrompt(
	evalCase: EvalCase,
	agentRun: AgentRun,
	outputFiles: Array<{ path: string; content: string }>,
): string {
	const turns = getTurns(evalCase);
	const isMultiTurn = turns.length > 1;

	const turnsSection = formatTurnsSection(turns, isMultiTurn);
	const transcriptText = formatTranscript(agentRun);
	const filesText = formatOutputFiles(outputFiles);
	const assertionsList = evalCase.assertions.map((a, i) => `${i + 1}. ${a}`).join("\n");

	const prompt = `You are an evaluator grading the output of an AI agent that was given a task.

## Original Task

${turnsSection}

**Expected behavior:**
${evalCase.expected_output}

## Assertions to Evaluate

Evaluate each of the following assertions. For each one, determine whether the agent's behavior satisfies it, and provide specific evidence from the transcript.

${assertionsList}

## Agent Transcript

${transcriptText}

## Output Files Produced

${filesText}

## Instructions

Return a JSON object with this exact structure (no other text, just JSON):

{
  "expectations": [
    {
      "text": "<the assertion text>",
      "passed": true/false,
      "evidence": "<specific evidence from the transcript>"
    }
  ],
  "claims": [],
  "eval_feedback": "<optional overall feedback or null>"
}

There must be exactly one entry in "expectations" for each assertion listed above, in the same order. Be strict: only mark an assertion as passed if the transcript clearly demonstrates it.`;

	return cap(prompt, MAX_GRADING_CHARS);
}

export const GRADER_SYSTEM_PROMPT =
	"You are a precise evaluator. You grade AI agent outputs against specific assertions. " +
	"Return only valid JSON matching the requested schema. Be strict but fair.";
